src/chaat/model/user.clj

```diff
@@ -0,0 +1,72 @@
+(ns chaat.model.user
+  (:require
+   [chaat.db.user :as db.user]
+   [crypto.password.bcrypt :as bcrypt]
+   [java-time.api :as jt]
+   [next.jdbc.date-time :as dt]
+   [chaat.errors :refer [do-or-error]]))
+
+;; Model layer does validation pertaining to the representation 
+;; of a user: username & password format (length, special characters etc.)
+
+(defn min-length?
+  "Check if string is >= a threshold length"
+  [str threshold]
+  (>= (count str) threshold))
+
+;; Add more restrictions: special characters not allowed, only alphanumeric etc.
+;; Perhaps I can start using the schema library? Or implement something similar myself.
+(defn validate-username-format
+  "Basic format check for username"
+  [username]
+  (let [min-length 2]
+    (if (min-length? username min-length)
+      {:result username :error nil}
+      {:result nil :error "Wrong username format"})))
+
+(defn validate-password-format
+  "Basic format check for password"
+  [password]
+  (let [min-length 8]
+    (if (min-length? password min-length)
+      {:result password :error nil}
+      {:result nil :error "Wrong password format"})))
+
+(defn validate-signup-details
+  "Basic format check for signup details: username & password"
+  [username password]
+  (let [result (validate-username-format username)
+        result (do-or-error result validate-password-format password)]
+    result))
+
+(defn get-time-instant
+  "Returns the result from a call to jt/instant"
+  []
+  (jt/instant))
+
+(defn gen-new-user-map
+  "Generate user info map for new user"
+  [username password]
+  (let [work-factor 11
+        password-hash (bcrypt/encrypt password work-factor)]
+    {:username username
+     :password_hash password-hash
+     :creation_timestamp (get-time-instant)
+     :display_picture nil}))
+
+;; can add optional parameters to help with testing
+;; can take an optional empty map
+(defn create
+  "Create a user and add user info to db"
+  [db username password]
+  (let [result (validate-signup-details username password)
+        user-info (do-or-error result gen-new-user-map username password)
```

Comment:
suggestion: can we move this call inside the next sexpr?

```
result (do-or-error result db.user/insert db (gen-new-user-map username password))
```

Reply:
maybe nto actually, this might cause errors. lets discuss

Reply:
I agree that we can shorten this. I have a macro now that can make this look much nicer, and will be updating this.

Reply:
resolving this for now.